src/main/java/com/hivemq/client/internal/mqtt/handler/publish/incoming/MqttSubscriptionFlowTree.java

```diff
@@ -145,47 +175,36 @@ void subscribe(final @Nullable MqttTopicLevel level, final @Nullable TopicTreeEn
                 }
                 multiLevelSubscriptions++;
             } else {
-                final TopicTreeNode node;
-                if (next == null) {
-                    next = new HashMap<>();
-                    node = null;
-                } else {
-                    node = next.get(level);
-                }
-                if (node == null) {
-                    if (level.isSingleLevelWildcard()) {
-                        hasSingleLevelSubscription = true;
+                if (entry != null) {
+                    if (entries == null) {
+                        entries = new HandleList<>();
                     }
-                    final ByteArray levelCopy = level.copy();
-                    next.put(levelCopy, new TopicTreeNode(levelCopy, level.next(), entry));
-                } else {
-                    node.subscribe(level.next(), entry);
+                    entries.add(entry);
                 }
+                subscriptions++;
             }
+            return null;
         }
 
-        boolean remove(final @Nullable MqttTopicLevel level, final @Nullable MqttSubscribedPublishFlow flow) {
-            if (level == null) {
-                if (remove(entries, flow)) {
-                    entries = null;
-                }
-                subscriptions--;
-                return (subscriptions == 0) && (multiLevelSubscriptions == 0) && (next == null);
+        @Nullable TopicTreeNode remove(
+                final @NotNull MqttTopicIterator topicIterator, final @Nullable MqttSubscribedPublishFlow flow) {
+
+            if (topicIterator.hasNext()) {
+                return traverseNext(topicIterator);
             }
-            if (level.isMultiLevelWildcard()) {
+            if (topicIterator.hasMultiLevelWildcard()) {
                 if (remove(multiLevelEntries, flow)) {
                     multiLevelEntries = null;
                 }
                 multiLevelSubscriptions--;
-                return (subscriptions == 0) && (multiLevelSubscriptions == 0) && (next == null);
-            }
-            if (next != null) {
-                final TopicTreeNode node = next.get(level);
-                if ((node != null) && node.remove(level.next(), flow)) {
-                    return removeNext(node);
+            } else {
+                if (remove(entries, flow)) {
+                    entries = null;
                 }
+                subscriptions--;
             }
-            return false;
+            compact();
+            return null;
```

Comment:
Does this always return null?
Could the return type be void?

Reply:
No, the first if branch may return not null